kong/tools/wrpc.lua

```diff
@@ -0,0 +1,686 @@
+require "table.new"
+local pb = require "pb"
+local semaphore = require "ngx.semaphore"
+local grpc = require "kong.tools.grpc"
+local channel = require "kong.tools.channel"
+
+local select = select
+local table_unpack = table.unpack     -- luacheck: ignore
+local table_insert = table.insert
+local table_remove = table.remove
+
+local exiting = ngx.worker.exiting
+
+local DEFAULT_EXPIRATION_DELAY = 90
+local CHANNEL_DICT_NAME = "wrpc_channel_dict"
+local CHANNEL_CLIENT_PREFIX = "wrpc_client_"
+
+pb.option("no_default_values")
+
+local wrpc = {}
+
+local function endswith(s, e) -- luacheck: ignore
+  return s and e and e ~= "" and s:sub(#s-#e+1, #s) == e
+end
+
+local Queue = {}
+Queue.__index = Queue
+
+function Queue.new()
+  return setmetatable({
+    smph = semaphore.new(),
+  }, Queue)
+end
+
+function Queue:push(itm)
+  table_insert(self, itm)
+  return self.smph:post()
+end
+
+function Queue:pop(timeout)
+  local ok, err = self.smph:wait(timeout or 1)
+  if not ok then
+    return nil, err
+  end
+
+  return table_remove(self, 1)
+end
+
+local semaphore_waiter
+do
+  local function handle(self, data)
+    self.data = data
+    self.smph:post()
+  end
+
+  local function handle_error(self, etype, errdesc)
+    self.data = nil
+    self.error = errdesc
+    self.etype = etype
+    self.smph:post()
+  end
+
+  local function expire(self)
+    self:handle_error("timeout", "timeout")
+  end
+
+  function semaphore_waiter()
+    return {
+      smph = semaphore.new(),
+      deadline = ngx.now() + DEFAULT_EXPIRATION_DELAY,
+      handle = handle,
+      handle_error = handle_error,
+      expire = expire,
+    }
+  end
+end
+
+
+local remote_waiter
+do
+  local function handle(self, payload)
+    channel.put_back(self.dict, self.name, pb.encode("wrpc.PayloadV1", payload))
+  end
+
+  local function handle_error(self, etype, errdesc)
+    channel.put_back(self.dict, self.name, pb.encode("wrpc.PayloadV1", {
+      mtype = "MESSAGE_TYPE_ERROR",
+      error = {
+        etype = etype,
+        description = errdesc,
+      }
+    }))
+  end
+
+  function remote_waiter(dict, name)
+    return {
+      dict = dict,
+      name = name,
+      deadline = ngx.now() + DEFAULT_EXPIRATION_DELAY,
+      handle = handle,
+      handle_error = handle_error,
+      raw = true,
+    }
+  end
+end
+
+
+local function merge(a, b)
+  if type(b) == "table" then
+    for k, v in pairs(b) do
+      a[k] = v
+    end
+  end
+
+  return a
+end
+
+local function proto_searchpath(name)
+  return package.searchpath(name, "kong/include/?.proto;/usr/include/?.proto")
+end
+
+--- definitions for the transport protocol
+local wrpc_proto
+
+
+local wrpc_service = {}
+wrpc_service.__index = wrpc_service
+
+
+--- a `service` object holds a set of methods defined
+--- in .proto files
+function wrpc.new_service()
+  if not wrpc_proto then
+    local wrpc_protofname = assert(proto_searchpath("wrpc.wrpc"))
+    wrpc_proto = assert(grpc.each_method(wrpc_protofname))
```

Comment:
It is a bit confusing that wRPC should call gRPC's method. Maybe we can extract the protobuf part out?